src/lib/support/StringCircularBuffer.cpp

```diff
@@ -0,0 +1,151 @@
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <support/StringCircularBuffer.h>
+
+#include <limits>
+#include <nlassert.h>
+#include <string.h>
+
+#include <support/CodeUtils.h>
+
+namespace chip {
+
+std::size_t StringCircularBuffer::Advance(std::size_t dataLocation, size_t amount) const
+{
+    dataLocation += amount;
+    if (dataLocation >= mCapacity)
+        dataLocation -= mCapacity;
+    return dataLocation;
+}
+
+void StringCircularBuffer::Read(uint8_t * dest, size_t length, size_t offset) const
+{
+    VerifyOrDie(Size() >= offset + length);
+
+    std::size_t start = Advance(mDataStart, offset);
+    if (mCapacity - start >= length)
+    {
+        ::memcpy(dest, mStorage + start, length);
+    }
+    else
+    {
+        size_t firstPiece  = mCapacity - start;
+        size_t secondPiece = length - firstPiece;
+        ::memcpy(dest, mStorage + start, firstPiece);
+        ::memcpy(dest + firstPiece, mStorage, secondPiece);
+    }
+}
+
+void StringCircularBuffer::Write(const uint8_t * source, size_t length)
+{
+    // Always reserve 1 byte to prevent mDataStart == mDataEnd.
+    VerifyOrDie(mCapacity - Size() - 1 >= length);
+
+    if (mCapacity - mDataEnd >= length)
+    {
+        ::memcpy(mStorage + mDataEnd, source, length);
+        mDataEnd = Advance(mDataEnd, length); // Use advance in case that mDataEnd wrap to 0
+    }
+    else
+    {
+        size_t firstPiece  = mCapacity - mDataEnd;
+        size_t secondPiece = length - firstPiece;
+        ::memcpy(mStorage + mDataEnd, source, firstPiece);
+        ::memcpy(mStorage, source + firstPiece, secondPiece);
+        mDataEnd = secondPiece;
+    }
+}
+
+void StringCircularBuffer::Drop(size_t length)
+{
+    VerifyOrDie(Size() >= length);
+    mDataStart = Advance(mDataStart, length);
+}
+
+size_t StringCircularBuffer::Size() const
+{
+    if (mDataStart <= mDataEnd)
+    {
+        return mDataEnd - mDataStart;
+    }
+    else
+    {
+        return mCapacity + mDataEnd - mDataStart;
+    }
+}
+
+CHIP_ERROR StringCircularBuffer::Push(const ByteSpan & payload)
+{
+    size_t length = payload.size();
+    if (length + sizeof(SizeType) + 1 > mCapacity)
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    if (length > std::numeric_limits<SizeType>::max())
+        return CHIP_ERROR_INVALID_ARGUMENT;
+
+    // Free up space until there is enough space.
+    while (length + sizeof(SizeType) + 1 > mCapacity - Size())
```

Comment:
Again, need to worry about overflow issues.  This time we can't just subtract off `sizeof(SizeType) + 1` because we don't know that we have that much space in our buffer....  So we probably need to loop doing `Pop` while `sizeof(SizeType) + 1 > mCapacity - Size()` and then do it again while `length > mCapacity - Size() - (sizeof(SizeType) + 1 )`.